posts/mathlib-changelog.md

```diff
@@ -0,0 +1,55 @@
+
+The rest of this post will talk about the engineering behind how this works.
+
+# How it works
+
+While building this, I had a few goals for the project:
+
+- It should be free to run
+- It should not require a backend
+- It should have a search capability
+- Google should be able to index all pages
+- It should stay up to date with mathlib automatically
+- It should be fully open-source
+
+To accomplish this, the project uses a GitHub repo as its main “database” containing the current state of the changelog in both JSON and TXT formats. The GitHub repo updates itself nightly on a cron using GitHub Actions to import new changes from mathlib.
+
+The main technical challenge of building the changelog is how to extract all the relevant changes from every commit to mathlib, and do it fast enough to run in a GitHub Action. Running a full Lean environment and parsing every commit, while it would be the most accurate (and awesome) solution, isn’t feasible because it’s too slow. mathlib has over 14,000 commits, and Lean takes 30+ seconds to load mathlib, meaning a faster solution was needed.
+
+The solution I settled on was to use Python and just scan through each modified file in each commit as a string and keep track of which tracked keywords are present. This isn’t guaranteed to catch everything, but it seems to work surprisingly well.
+
+The website itself is a static site built using [NextJS](https://nextjs.org/) and generated from the JSON version of the changelog in the GitHub repo. NextJS has the added benefit of being able to lazily generate static pages. There are over 100,000 pages on the changelog, so generating and uploading them each up-front would take too long to be feasible in a GitHub Action (I tried).
+
+The website search is all handled on the frontend. It just downloads a full list of every item in the changelog and does full-text search locally in the browser. It’s not the most advanced search in the world, but being run locally at least means it’s extremely fast!
+
+If there’s anything the changelog is missing, or if you have any ideas for improvements, open an issue or pull request in the [GitHub repo](https://github.com/chanind/mathlib-changelog). Contributions are welcome!
+
+# Thank you to the mathlib community!
+
+Thank you to everyone on Zulip who tried out the changelog, found bugs, offered suggestions, and helped brainstorm ideas for how this can work. This includes Alex Best, Damiano Testa, Eric Wieser, Floris van Doorn, Heather Macbeth, Johan Commelin, Kevin Buzzard, Kyle Miller, and others!
```

Comment:
Could insert a link to the Zulip thread here if you wanted too.
```suggestion
Thank you to everyone [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/113489-new-members/topic/mathlib.20versioning.20and.20changelog/near/288918851) who tried out the changelog, found bugs, offered suggestions, and helped brainstorm ideas for how this can work. This includes Alex Best, Damiano Testa, Eric Wieser, Floris van Doorn, Heather Macbeth, Johan Commelin, Kevin Buzzard, Kyle Miller, and others!
```

Up to you whether you think that's a good idea or not.

Reply:
I'd vote to leave it out since some people also sent private messages on Zulip with feedback rather than posting in the chat. I don't feel strongly though, so if others want to link it that's OK too

Reply:
Fine by me